File: leetcode/112-path-sum.rb
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @param {Integer} sum
# @return {Boolean}
def has_path_sum(root, sum)
  if root.nil?
    false
  elsif root.left.nil? and root.right.nil?
    sum == root.val
  else
    left = has_path_sum root.right, sum - root.val
    left or (has_path_sum root.left, sum - root.val)
  end
end

File: leetcode/112-path-sum.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode *root, int sum) {
        if (root == nullptr) {
            return false;
        }
        if (sum == root->val && root->left == nullptr && root->right == nullptr) {
            return true;
        }
        return (root->left != nullptr && hasPathSum(root->left, sum - root->val))
            || (root->right != nullptr && hasPathSum(root->right, sum - root->val));
    }
};
